### PR TITLE
Add CSV,XML parser.

### PR DESCRIPTION
**Pull request by p4535992**
Add csv parser.
A functional example here:
http://jsfiddle.net/4535992/gynog5nx/2/

### CODE REVIEW

**Comment by leplatrem** [comment]
Thanks for this contribution!

Let me give you some feedback before it could be merged:
* Add a small test. at least to check that providing a small piece of csv it comes out as some geojson
* Document the options you added in the readme (*mentioning when required, with default value*)
* Add safety checks in case the dependency you introduced was not provided
* Add yourself to the authors

If you update the pull-request, please let us know with a small comment otherwise I won't be notified!

Thanks again!

File: leaflet.filelayer.js [hunk 1]
```diff
@@ -9,7 +9,11 @@ var FileLoader = L.Class.extend({
     includes: L.Mixin.Events,
     options: {
         layerOptions: {},
-        fileSizeLimit: 1024
+        fileSizeLimit: 1024,
+        firstLineTitles: true,
```
Comment (leplatrem) [inline leaflet.filelayer.js]: I believe this should be name *headers*

File: leaflet.filelayer.js [hunk 2]
```diff
@@ -9,7 +9,11 @@ var FileLoader = L.Class.extend({
...
+        latitudeTitle: 'lat',
+        longitudeTitle: 'lng',
```
Comment (leplatrem) [inline leaflet.filelayer.js]: what about *column* instead of *title*?

File: leaflet.filelayer.js [hunk 3]
```diff
@@ -17,9 +21,11 @@ var FileLoader = L.Class.extend({
         L.Util.setOptions(this, options);
 
         this._parsers = {
+            'json': this._loadGeoJSON,
```
Comment (leplatrem) [inline leaflet.filelayer.js]: :+1:

File: leaflet.filelayer.js [hunk 4]
```diff
@@ -82,13 +88,81 @@ var FileLoader = L.Class.extend({
...
+        }
+        return papaJson;
+    },
+
+    _deleteDoubleQuotes: function (text) {
+        text = text.trim().replace(/^"/,"").replace(/"$/,"");
+        return text;
+    },
+
+    _propertiesNames: []
```
Comment (leplatrem) [inline leaflet.filelayer.js]: This seems unused.

File: leaflet.filelayer.js [hunk 5]
```diff
@@ -82,13 +88,81 @@ var FileLoader = L.Class.extend({
...
+            if(!this.options.firstLineTitles){
+                throw new Error('The file CSV must have the Headers');
+            }
+            var geojson;
+            /*
+             * for this function i used the Papa Parser of mholt
+             * (https://github.com/mholt/PapaParse)
+             */
+            geojson = Papa.parse(content,{header: this.options.firstLineTitles}); //convert csv to json.
+            geojson = this._addFeatureToJson(geojson);
```
Comment (leplatrem) [inline leaflet.filelayer.js]: This function could be named `papaJsonToGeoJSON` instead

File: leaflet.filelayer.js [hunk 6]
```diff
@@ -82,13 +88,81 @@ var FileLoader = L.Class.extend({
...
+            return this._loadGeoJSON(geojson);
+        }catch(e){alert(e.message);}
```
Comment (leplatrem) [inline leaflet.filelayer.js]: Prefer `this.fire('data:error', {error: e});` instead

File: leaflet.filelayer.js [hunk 7]
```diff
@@ -9,7 +9,11 @@ var FileLoader = L.Class.extend({
     includes: L.Mixin.Events,
     options: {
         layerOptions: {},
-        fileSizeLimit: 1024
+        fileSizeLimit: 1024,
+        firstLineTitles: true,
+        latitudeTitle: 'lat',
+        longitudeTitle: 'lng',
+        titleForSearch: 'title'
```
Comment (leplatrem) [inline leaflet.filelayer.js]: What is this?

File: leaflet.filelayer.js [hunk 8]
```diff
@@ -82,13 +88,81 @@ var FileLoader = L.Class.extend({
...
+                            if (obj[title[i]].indexOf('http') === 0) {
+                                href = '<a target="_blank" href="' + obj[title[i]] + '">'+ obj[title[i]] + '</a>';
+                            }
+                            if(href.length > 0)content += '<tr><th>'+title[i]+'</th><td>'+ href +'</td></tr>';
+                            else content += '<tr><th>'+title[i]+'</th><td>'+ obj[title[i]] +'</td></tr>';
+                        }
+                    }//end of if
+                }//end of for
+                content += "</table></div>";
+                feature["properties"]["popupContent"]=content;
```
Comment (leplatrem) [inline leaflet.filelayer.js]: This whole popContent seems specific to your use case.
You should put every column in `properties`, and let the users decide what they do with them (html popup or whatever).

File: leaflet.filelayer.js [hunk 9]
```diff
@@ -82,13 +88,81 @@ var FileLoader = L.Class.extend({
...
+            var fields = title.toString().trim().split(",");
+            var lng = parseFloat(obj[this.options.longitudeTitle]);
+            var lat = parseFloat(obj[this.options.latitudeTitle]);
+            if (fields.length+"=="+title.length && lng<180 && lng>-180 && lat<90 && lat>-90) {
+                var feature = {};
+                feature["type"]="Feature";
+                feature["geometry"]={};
+                feature["properties"]={};
+                feature["geometry"]["type"]="Point";
+                feature["geometry"]["coordinates"]=[lng,lat];
```
Comment (leplatrem) [inline leaflet.filelayer.js]: Could be done in one line `feature["geometry"] = {"type": "Point", "coordinates": [lng, lat]}`

File: leaflet.filelayer.js [hunk 10]
```diff
@@ -82,13 +88,81 @@ var FileLoader = L.Class.extend({
...
+                var content ='<div class="popup-content"><table class="table table-striped table-bordered table-condensed">';
+                for (var i=0; i< title.length; i++) {
+                    if (title[i] != this.options.latitudeTitle && title[i] != this.options.longitudeTitle) {
+                        if(title[i]==this.options.titleForSearch){
+                            feature["properties"]["search"]=this._deleteDoubleQuotes(obj[title[i]]);
+                            feature["properties"]["title"]=this._deleteDoubleQuotes(obj[title[i]]);
+                        }else{
```
Comment (leplatrem) [inline leaflet.filelayer.js]: Please respect the formatting of the rest of the file (also for variable assignation etc.)

File: leaflet.filelayer.js [hunk 11]
```diff
@@ -82,13 +88,81 @@ var FileLoader = L.Class.extend({
...
+    _addFeatureToJson: function(papaJson){
+        papaJson["type"]="FeatureCollection";
+        papaJson["features"]=[];
+        var title = papaJson.meta.fields;
+        for (var num_linea = 0; num_linea < papaJson.data.length - 1; num_linea++) { //  var depth = papaJson.data.length - 1;
+            var obj = papaJson.data[num_linea]; //single element of papa parse json object
+            var fields = title.toString().trim().split(",");
+            var lng = parseFloat(obj[this.options.longitudeTitle]);
+            var lat = parseFloat(obj[this.options.latitudeTitle]);
+            if (fields.length+"=="+title.length && lng<180 && lng>-180 && lat<90 && lat>-90) {
```
Comment (leplatrem) [inline leaflet.filelayer.js]: What is `fields.length+"=="+title.length` ?

Comment (leplatrem) [inline leaflet.filelayer.js]: Instead you should probably do something like:

```js
if (!(conditions)) {
    console.warn("Ignore line", num_linea, " invalid data");
    continue;
}
```

File: leaflet.filelayer.js [hunk 12]
```diff
@@ -82,13 +88,81 @@ var FileLoader = L.Class.extend({
...
+                            else content += '<tr><th>'+title[i]+'</th><td>'+ obj[title[i]] +'</td></tr>';
+                        }
+                    }//end of if
+                }//end of for
+                content += "</table></div>";
+                feature["properties"]["popupContent"]=content;
+                papaJson["features"].push(feature);
+            }
+        }
+        return papaJson;
```
Comment (leplatrem) [inline leaflet.filelayer.js]: What you return here is not papaJson, but a geoJSON collection

**Comment by p4535992** [comment]
Sorry if I did not answer right away, thank you for the advice, I will make the appropriate changes and let you know.

**Comment by p4535992** [comment]
Add a new commit, with a xml parser, and gtfs viewer (need some upgrade aniway) .

**Comment by mabhub** [comment]
I don't think adding parsers is the best way to achieve this, but maybe adding a way to use middleware or hooks would be better.

Closing for now, but please comment if needed.

**Pull request closed without merging**